<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2021-08-08" />

<title>Day 2 - Module 09: Sample Visualizations for Quality Control</title>

<script src="site_libs/header-attrs-2.8/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02_QC.html">Quality Control</a>
    </li>
    <li>
      <a href="Module02optional_Cutadapt_MultiQC.html">Optional: Cutadapt/MultiQC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module04_Alignment.html">Alignment/Quantification (STAR/RSEM)</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module05_MultiQC_and_Count_Matrix.html">Count matrix / Alignment QC</a>
    </li>
    <li>
      <a href="Module05optional_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">Introduction to DESeq2</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DESeq2 Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">Count Normalizations and DESeq2 model fitting</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample Visualizations for QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Comparisons and Gene Annotations</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Day 2 - Module 09: Sample Visualizations for Quality Control</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2021-08-08</h4>

</div>


<!--- Allow the page to be wider --->
<style>
    body .main-container {
        max-width: 1200px;
    }
</style>
<blockquote>
<h1 id="objectives">Objectives:</h1>
<ul>
<li>Generate common QC visualizations<br />
</li>
<li>Understand how to interpret QC visualizations<br />
</li>
<li>Understand when to revise the model used in the DESeq2 initialization<br />
</li>
<li>Understand the pitfalls of post-hoc analysis<br />
</li>
<li>Describe the causes and implications of batch effect or other QC issues in an RNA-Seq experiment</li>
</ul>
</blockquote>
<div id="differential-expression-workflow" class="section level1" number="1">
<h1 number="1"><span class="header-section-number">1</span> Differential Expression Workflow</h1>
<p>Prior to testing for differential expression between our comparisons of interest, we’ll first generate plots that will assess how well our samples match up with our expectations (based on their treatment groups) and what we might expect to see from our differential expression comparisons.</p>
<table>
<thead>
<tr class="header">
<th align="center">Step</th>
<th>Task</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">1</td>
<td>Experimental Design</td>
</tr>
<tr class="even">
<td align="center">2</td>
<td>Biological Samples / Library Preparation</td>
</tr>
<tr class="odd">
<td align="center">3</td>
<td>Sequence Reads</td>
</tr>
<tr class="even">
<td align="center">4</td>
<td>Assess Quality of Raw Reads</td>
</tr>
<tr class="odd">
<td align="center">5</td>
<td>Splice-aware Mapping to Genome</td>
</tr>
<tr class="even">
<td align="center">6</td>
<td>Count Reads Associated with Genes</td>
</tr>
<tr class="odd">
<td align="center">—</td>
<td>—</td>
</tr>
<tr class="even">
<td align="center">7</td>
<td>Organize project files locally</td>
</tr>
<tr class="odd">
<td align="center">8</td>
<td>Initialize DESeq2 and fit DESeq2 model</td>
</tr>
<tr class="even">
<td align="center"><strong>9</strong></td>
<td><strong>Assess expression variance within treatment groups</strong></td>
</tr>
<tr class="odd">
<td align="center">10</td>
<td>Specify pairwise comparisons and test for differential expression</td>
</tr>
<tr class="even">
<td align="center">11</td>
<td>Generate summary figures for comparisons</td>
</tr>
<tr class="odd">
<td align="center">12</td>
<td>Annotate differential expression result tables</td>
</tr>
</tbody>
</table>
<hr />
</div>
<div id="sample-visualizatons-for-quality-control" class="section level1" number="2">
<h1 number="2"><span class="header-section-number">2</span> Sample Visualizatons for Quality Control</h1>
<p>Yesterday we discussed aspects of quality control assessment at the sequencing data level. Today we will outline sample-level and gene level quality control to determine what we should expect from our downstream differential expression comparisons.</p>
<p>To do this, we will assess the similarity of our samples by using principal component analysis (PCA). This will allow us to determine how well patterns in the data fits our expectations from the experiments design and possible sources of variation in the dataset.</p>
<p>Other common visualizations that we generate for our analyses include expression heatmaps, sample correlation heatmaps, and boxplots of raw and/or normalized counts.</p>
<p><strong>TODO: Link to bonus content which includes code for additional visualizations</strong></p>
<div id="principle-component-analysis-pca-plots" class="section level2" number="2.1">
<h2 number="2.1"><span class="header-section-number">2.1</span> Principle Component Analysis (PCA) Plots</h2>
<p>A common and very useful plot for evaluating how well our samples cluster by treatment groups are Principle Component Analysis (PCA) plots. PCA is used to emphasize variation and bring out patterns in large datasets by using dimensionality redution.</p>
<p>This image from <a href="https://blog.bioturing.com/2018/06/14/principal-component-analysis-explained-simply/">a helpful step by step explaination of PCA</a> helps to illustrate the principle component projections for two genes measured in approximately 60 mouse samples. Generally, this process is repeated and after each gene’s contribution to a principle component or weight is determined, the expression and weight are summed across genes for each sample to calculate a value for each principle component.</p>
<p><img src="images/Blog_pca_6b.png" /></p>
<blockquote>
<p><strong>Note</strong>: A more detailed overview of the PCA procedure is outlined in <a href="https://hbctraining.github.io/DGE_workshop/lessons/principal_component_analysis.html">a Harvard Chan Bioinformatic Core training module</a> and is based a more thorough description presented in a <a href="https://www.youtube.com/watch?v=_UVHneBUBW0">StatQuest’s video</a>. Additionally, <a href="https://towardsdatascience.com/principal-component-analysis-3c39fbf5cb9d">this TowardsDataScience blog post</a> goes through the math behind PCAs.</p>
</blockquote>
<div id="interpreting-pca-plots" class="section level3" number="2.1.1">
<h3 number="2.1.1"><span class="header-section-number">2.1.1</span> Interpreting PCA plots</h3>
<p><strong>TODO: Add image of PCA to walk through interpretation before going through code</strong></p>
<p>In the plot we’ll generate below, we can see that principle component 1 (PC1) explains ~80% of the variance in our data while principle component 2 (PC2) explains ~12% of the variance. We also see that samples in both <code>control</code> groups are fairly tightly grouped, while samples within each <code>Tx</code> group do not cluster as tightly.</p>
<p>We generally expect most of the variance to be explained by the first two or three principle components.</p>
<p>This <a href="https://blog.bioturing.com/2018/06/14/principal-component-analysis-explained-simply/">helpful overview of PCA basics</a> walks through both the generation and interpretatation of similar plots.</p>
</div>
<div id="evaulating-batch-effects-or-confounders" class="section level3" number="2.1.2">
<h3 number="2.1.2"><span class="header-section-number">2.1.2</span> Evaulating batch effects or confounders</h3>
<p>PCA plots are useful for evaulating the impact of factors other than the experimental treatment or group.</p>
<p>At times, batch effects can be quite obvious, such as this example from the <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">DESeq2 vignette</a>, where there is clear separation within the treatment groups.</p>
<p><img src="images/PCA1_DESeq2Vignette.png" /></p>
<p>If we color only by sequencing run type (paired-end vs. single-end), we see that PC2 (29% of variance) is primarily explained by this technical covariate.</p>
<p><img src="images/PCA1_DESeq2Vignette.png" /></p>
<p>However, the samples are clearly seperated by experimental condition on PC1, and since PC1 explains more variance than PC2 <strong>and</strong> since we have non-confounded batches, we could incorporate the technical covariate into our model design, such as outlined in the <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs">DESeq2 vignette</a>.</p>
<details>
<summary>
<em>Click for complex design discussion</em>
</summary>
In experiments with more complex designs, such as when there are interesecting/multiple treatment conditions, it can be less clear what covariants are influencing expression, such as illustrated from <a href="http://www.molmine.com/magma/global_analysis/batch_effect.html">this documenation for a microarray analysis tool</a>. From the PCA labeled by experimental treatment, we see that samples from the treatment group do not cluster together and that there is high variance across all treatment groups. <img src="images/batch_ex1b.jpg" /> However, when the plot is color coded by the technical batches of probe labeling, we see that the patterns in the data are better explained by batch than the experimental conditions. <img src="images/batch_ex1c.jpg" />
</details>
<div id="when-to-remove-samples-or-update-the-design-formula" class="section level4" number="2.1.2.1">
<h4 number="2.1.2.1"><span class="header-section-number">2.1.2.1</span> When to remove samples or update the design formula</h4>
<p>Generally, the decision to remove samples should not be taken lightly, especially if the consideration is based on the expression level patterns we are visualizing in this section, as doing so without sufficient justification could potentially bias your results. Other training materials have a step-by-step example of evaluating <a href="https://hbctraining.github.io/DGE_workshop/lessons/03_DGE_QC_analysis.html">batch effects</a> using PCAs, that we would recommend reviewing prior to removing any samples from your comparisons.</p>
<p>If your PCA reveals patterns that were not accounted for in the initial model fit, the <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs">DESeq2 vignette</a> provides an overview of how to add additional covariates to a model design before refitting the DESeq2 model for a dataset.</p>
</div>
</div>
</div>
<div id="plot-setup" class="section level2" number="2.2">
<h2 number="2.2"><span class="header-section-number">2.2</span> Plot Setup</h2>
<p>We’ve already loaded the libraries we need for this module. We’ll follow best practices and create new directories to organize our output figures as well as a variable to store a descriptor of the dataset we are looking at.</p>
<pre class="r"><code>plotPath = &#39;figures/&#39;
dir.create(plotPath, showWarnings = FALSE)
Comparison &lt;- &quot;ko.Tx&quot; # descriptor for the dataset we are looking at
# NOTE, is this correct? In the text below &#39;Gtype.Tx&#39; is mentioned...</code></pre>
</div>
<div id="pca-plot-initialization" class="section level2" number="2.3">
<h2 number="2.3"><span class="header-section-number">2.3</span> PCA Plot initialization</h2>
<p>Below, we will plot the rlog normalized data for our samples projected onto a 2D plane and spanned by their first two principle components to visualize the overall effect of experimental covariates and determine if there is evidence of batch effects.</p>
<p>We’ll generate the PCA projections for the top 500 using the <code>plotPCA</code> function from DESeq2, specifying <code>Gtype.Tx</code> as the condition of interest.</p>
<pre class="r"><code>p.all &lt;- plotPCA(rld, intgroup = c(&#39;Gtype.Tx&#39;), ntop = 500)
p.all</code></pre>
<p><img src="Module09_SampleQCViz_files/figure-html/PCArlog3-1.png" width="672" /></p>
<p>Next, we’ll save this plot to file.</p>
<pre class="r"><code>pdf(file = paste0(plotPath, &#39;PCAplot_rlog_&#39;, Comparison, &#39;.pdf&#39;), onefile = TRUE)
p.all
dev.off()</code></pre>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<details>
<summary>
<em>Click for code to customize PCA plots</em>
</summary>
<p>We can generate a customized plot using the <code>ggplot2</code> package, including shape assignments for our treatment groups and color assignments for the individual samples.</p>
<pre class="r"><code>pdf(file = paste0(plotPath, &#39;PCAplot_Fancy_rlog_&#39;, Comparison, &#39;.pdf&#39;), onefile = TRUE)

#PCA plot for Rlog-Normalized counts for all samples
CombinatoricGroup &lt;- factor(MetaInfo$Gtype.Tx)
SampleName &lt;- factor(row.names(MetaInfo))


#Generate the PCA projections using the `plotPCA` function from DESeq2.
p.all &lt;- plotPCA(rld, intgroup = c(&#39;Gtype.Tx&#39;), ntop = 500)

gp &lt;- ggplot(p.all$data, aes(x = PC1, y = PC2, color = SampleName, shape = CombinatoricGroup)) +
    scale_shape_manual(values=1:nlevels(CombinatoricGroup), name = &#39;Combinatoric Group&#39;) +
    geom_point(size=2) +
    ggtitle(label = as.character(&#39;All samples Rlog-Normalized&#39;)) +
    guides(colour=guide_legend(nrow=12, title = &#39;Sample&#39;),
        legend.key = element_rect(size = 1),
        legend.key.size = unit(0, &#39;cm&#39;)) +
    xlab(p.all$labels[2]) + ylab(p.all$labels[1]) +
    theme(legend.margin = margin(t = 0, unit=&#39;mm&#39;), plot.title = element_text(hjust = 0.5)) +
    theme_classic(base_size = 10)
plot(gp)
dev.off()</code></pre>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
</details>
<details>
<summary>
<em>Click for example code for generating a ScreePlot</em>
</summary>
<p>A screeplot is a way to visualize the variance explained by all principle components. To generate a scree plot, the PCA results need to be used independently of plotting, such as described by <a href="https://statquest.org/pca-clearly-explained/">this statquest post</a> and replicated below.</p>
<pre class="r"><code>    # generate PCA loadings
    pca &lt;- prcomp(t(assay(rld)), scale=TRUE)
    
    ## get the scree information
    pca.var &lt;- pca$sdev^2
    scree &lt;- pca.var/sum(pca.var)
    barplot((scree[1:10]*100), main=&quot;Scree Plot&quot;, xlab=&quot;Principal Component&quot;, ylab=&quot;Percent Variation&quot;) 
    ```
    
    &lt;img src=&quot;Module09_SampleQCViz_files/figure-html/ScreePlot-1.png&quot; width=&quot;672&quot; /&gt;
    We can see that the majority (~65%) of the variance across our samples is explained by the first three principle components, giving us some additional confidence regarding the quality of our data.
    In these scree plot examples from BioTuring, the plot on the left fits what we would expect for a dataset with high signal from the experimental treatment, where the majority of the variance is explained by the first few principle components. The plot on the right illustrates a scenario where the variance is distributed across many components, which could be due to low signal from the experimental treatment, complex experimental design, or confounding factors.
image: ![](./images/proportion-of-variance-blog-horz.jpg)
&lt;/details&gt;


#### Optional PCA plot for raw data
&lt;details&gt;
    &lt;summary&gt;*Click for code for PCA visualization of raw data*&lt;/summary&gt;
    It can sometimes be useful to also generate a PCA plot for the raw data as well as the normalized data.
    
    ```r
    pdf(file = paste0(plotPath, &#39;PCAplot_raw_&#39;, Comparison, &#39;.pdf&#39;), onefile = TRUE)
    #PCA for Raw counts for all samples
    RC &lt;- SummarizedExperiment(log2(counts(dds, normalized = FALSE)), colData=colData(dds))
    p.RC &lt;- plotPCA(DESeqTransform(RC), intgroup = &#39;Gtype.Tx&#39;)
    
    gpRC &lt;- ggplot(p.RC$data, aes(x = PC1, y = PC2, color = SampleName, shape = CombinatoricGroup)) +     xlab(p.RC$labels[2]) + ylab(p.RC$labels[1]) + scale_shape_manual(values=1:nlevels(CombinatoricGroup), name = &#39;Combinatoric Group&#39;) + geom_point(size=2) + ggtitle(label = as.character(&#39;All samples Raw Data&#39;)) + theme(plot.title = element_text(hjust = 0.5)) + guides(colour=guide_legend(nrow=12, title = &#39;Sample&#39;), legend.key = element_rect(size = 1), legend.key.size = unit(0, &#39;cm&#39;)) + theme_classic(base_size = 10) + theme(legend.margin=margin(t = 0, unit=&#39;mm&#39;))
    plot(gpRC)
    
    dev.off()
    ```
    
    ```
    ## quartz_off_screen 
    ##                 2
    ```
    
    ```r
    # embedd example of plot (rlog only)
    plot(gpRC)
    ```
    
    &lt;img src=&quot;Module09_SampleQCViz_files/figure-html/PCAraw-1.png&quot; width=&quot;672&quot; /&gt;
    We see that there is less variance explained by PC1 and that the samples from the same group are not as well clustered for the raw data. Since this is prior to normalization, these differences are likely due to **technical** considerations like sequencing depth differences that are accounted for by the rlog normalization.   
&lt;/details&gt;

**TODO: Add in summary from last workshop materials**



---


# Sources Used    
* HBC QC tutorial: https://hbctraining.github.io/DGE_workshop/lessons/03_DGE_QC_analysis.html    
* Detailed Heatmap tutorial from Galaxy: https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-heatmap2/tutorial.html   
* PCA Overview: https://blog.bioturing.com/2018/06/14/principal-component-analysis-explained-simply/     






---

# Session Info:

```r
sessionInfo()</code></pre>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 11.5.1
## 
## Matrix products: default
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] dplyr_1.0.6                 data.table_1.14.0           RColorBrewer_1.1-2          pheatmap_1.0.12            
##  [5] ggrepel_0.9.1               tidyr_1.1.3                 ggplot2_3.3.4               DESeq2_1.32.0              
##  [9] SummarizedExperiment_1.22.0 Biobase_2.30.0              MatrixGenerics_1.4.0        matrixStats_0.58.0         
## [13] GenomicRanges_1.44.0        GenomeInfoDb_1.28.0         IRanges_2.26.0              S4Vectors_0.30.0           
## [17] BiocGenerics_0.38.0         rmarkdown_2.8              
## 
## loaded via a namespace (and not attached):
##  [1] sass_0.4.0             bit64_4.0.5            jsonlite_1.7.2         splines_4.1.0          bslib_0.2.5.1         
##  [6] assertthat_0.2.1       highr_0.9              blob_1.2.1             GenomeInfoDbData_1.2.6 yaml_2.2.1            
## [11] pillar_1.6.1           RSQLite_2.2.7          lattice_0.20-44        glue_1.4.2             digest_0.6.27         
## [16] XVector_0.32.0         colorspace_2.0-1       htmltools_0.5.1.1      Matrix_1.3-4           XML_3.99-0.6          
## [21] pkgconfig_2.0.3        genefilter_1.74.0      zlibbioc_1.16.0        purrr_0.3.4            xtable_1.8-4          
## [26] scales_1.1.1           BiocParallel_1.26.0    tibble_3.1.2           annotate_1.48.0        farver_2.1.0          
## [31] generics_0.1.0         ellipsis_0.3.2         withr_2.4.2            cachem_1.0.5           survival_3.2-11       
## [36] magrittr_2.0.1         crayon_1.4.1           memoise_2.0.0          evaluate_0.14          fansi_0.5.0           
## [41] tools_4.1.0            lifecycle_1.0.0        stringr_1.4.0          locfit_1.5-9.4         munsell_0.5.0         
## [46] DelayedArray_0.18.0    AnnotationDbi_1.32.3   compiler_4.1.0         jquerylib_0.1.4        rlang_0.4.11          
## [51] grid_4.1.0             RCurl_1.98-1.3         labeling_0.4.2         bitops_1.0-7           gtable_0.3.0          
## [56] DBI_1.1.1              R6_2.5.0               knitr_1.33             fastmap_1.1.0          bit_4.0.4             
## [61] utf8_1.2.1             stringi_1.6.2          Rcpp_1.0.6             vctrs_0.3.8            geneplotter_1.70.0    
## [66] tidyselect_1.1.1       xfun_0.23</code></pre>
<hr />
<p>These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the <a href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.</p>
</div>
</div>

<div id="rmd-source-code">---
title: "Day 2 - Module 09: Sample Visualizations for Quality Control"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: true
            fig_caption: true
            markdown: GFM
            code_download: true
---

<!--- Allow the page to be wider --->
<style>
    body .main-container {
        max-width: 1200px;
    }
</style>

> # Objectives:     
> * Generate common QC visualizations   
> * Understand how to interpret QC visualizations    
> * Understand when to revise the model used in the DESeq2 initialization    
> * Understand the pitfalls of post-hoc analysis     
> * Describe the causes and implications of batch effect or other QC issues in an RNA-Seq experiment     


```{r Modules, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
library(ggplot2)
library(tidyr)
library(dplyr)
library(matrixStats)
library(ggrepel)
library(pheatmap)
library(RColorBrewer)
load("rdata/RunningData.RData")
```


# Differential Expression Workflow

Prior to testing for differential expression between our comparisons of interest, we'll first generate plots that will assess how well our samples match up with our expectations (based on their treatment groups) and what we might expect to see from our differential expression comparisons.

| Step | Task |
| :--: | ---- |
| 1 | Experimental Design |
| 2 | Biological Samples / Library Preparation |
| 3 | Sequence Reads |
| 4 | Assess Quality of Raw Reads |
| 5 | Splice-aware Mapping to Genome |
| 6 | Count Reads Associated with Genes |
| --- | --- |
| 7 | Organize project files locally |
| 8 | Initialize DESeq2 and fit DESeq2 model |
| **9** | **Assess expression variance within treatment groups** |
| 10 | Specify pairwise comparisons and test for differential expression |
| 11 | Generate summary figures for comparisons |
| 12 | Annotate differential expression result tables |

---

# Sample Visualizatons for Quality Control     

Yesterday we discussed aspects of quality control assessment at the sequencing data level. Today we will outline sample-level and gene level quality control to determine what we should expect from our downstream differential expression comparisons.

To do this, we will assess the similarity of our samples by using principal component analysis (PCA). This will allow us to determine how well patterns in the data fits our expectations from the experiments design and possible sources of variation in the dataset.

Other common visualizations that we generate for our analyses include expression heatmaps, sample correlation heatmaps, and boxplots of raw and/or normalized counts. 

**TODO: Link to bonus content which includes code for additional visualizations**

## Principle Component Analysis (PCA) Plots

A common and very useful plot for evaluating how well our samples cluster by treatment groups are Principle Component Analysis (PCA) plots. PCA is used to emphasize variation and bring out patterns in large datasets by using dimensionality redution. 

This image from
[a helpful step by step explaination of PCA](https://blog.bioturing.com/2018/06/14/principal-component-analysis-explained-simply/) helps to illustrate the principle component projections for two genes measured in approximately 60 mouse samples. Generally, this process is repeated and after each gene's contribution to a principle component or weight is determined, the expression and weight are summed across genes for each sample to calculate a value for each principle component. 


![](./images/Blog_pca_6b.png)

>**Note**: A more detailed overview of the PCA procedure is outlined in [a Harvard Chan Bioinformatic Core training module](https://hbctraining.github.io/DGE_workshop/lessons/principal_component_analysis.html) and is based a more thorough description presented in a [StatQuest’s video](https://www.youtube.com/watch?v=_UVHneBUBW0). Additionally, [this TowardsDataScience blog post](https://towardsdatascience.com/principal-component-analysis-3c39fbf5cb9d) goes through the math behind PCAs.

### Interpreting PCA plots

**TODO: Add image of PCA to walk through interpretation before going through code**

In the plot we'll generate below, we can see that principle component 1 (PC1) explains ~80% of the variance in our data while principle component 2 (PC2) explains ~12% of the variance. We also see that samples in both `control` groups are fairly tightly grouped, while samples within each `Tx` group do not cluster as tightly. 

We generally expect most of the variance to be explained by the first two or three principle components.


This [helpful overview of PCA basics](https://blog.bioturing.com/2018/06/14/principal-component-analysis-explained-simply/) walks through both the generation and interpretatation of similar plots.    


### Evaulating batch effects or confounders

PCA plots are useful for evaulating the impact of factors other than the experimental treatment or group. 

At times, batch effects can be quite obvious, such as this example from the [DESeq2 vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html), where there is clear separation within the treatment groups. 


![](./images/PCA1_DESeq2Vignette.png)

If we color only by sequencing run type (paired-end vs. single-end), we see that PC2 (29% of variance) is primarily explained by this technical covariate. 

![](./images/PCA1_DESeq2Vignette.png)

However, the samples are clearly seperated by experimental condition on PC1, and since PC1 explains more variance than  PC2 **and** since we have non-confounded batches, we could incorporate the technical covariate into our model design, such as outlined in the [DESeq2 vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs).

<details>
    <summary>*Click for complex design discussion*</summary>
    In experiments with more complex designs, such as when there are interesecting/multiple treatment conditions, it can be less clear what covariants are influencing expression, such as illustrated from [this documenation for a microarray analysis tool](http://www.molmine.com/magma/global_analysis/batch_effect.html).
    From the PCA labeled by experimental treatment, we see that samples from the treatment group do not cluster together and that there is high variance across all treatment groups.
    ![](./images/batch_ex1b.jpg)
    However, when the plot is color coded by the technical batches of probe labeling, we see that the patterns in the data are better explained by batch than the experimental conditions.
    ![](./images/batch_ex1c.jpg)
</details>

#### When to remove samples or update the design formula

Generally, the decision to remove samples should not be taken lightly, especially if the consideration is based on the expression level patterns we are visualizing in this section, as doing so without sufficient justification could potentially bias your results. Other training materials have a step-by-step example of evaluating [batch effects](https://hbctraining.github.io/DGE_workshop/lessons/03_DGE_QC_analysis.html) using PCAs, that we would recommend reviewing prior to removing any samples from your comparisons. 


If your PCA reveals patterns that were not accounted for in the initial model fit, the [DESeq2 vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs) provides an overview of how to add additional covariates to a model design before refitting the DESeq2 model for a dataset.

## Plot Setup

We've already loaded the libraries we need for this module. We'll follow best practices and create new directories to organize our output figures as well as a variable to store a descriptor of the dataset we are looking at.

```{r Setupdirectory}
plotPath = 'figures/'
dir.create(plotPath, showWarnings = FALSE)
Comparison <- "ko.Tx" # descriptor for the dataset we are looking at
# NOTE, is this correct? In the text below 'Gtype.Tx' is mentioned...
```

## PCA Plot initialization

Below, we will plot the rlog normalized data for our samples projected onto a 2D plane and spanned by their first two principle components to visualize the overall effect of experimental covariates and determine if there is evidence of batch effects. 

We'll generate the PCA projections for the top 500 using the `plotPCA` function from DESeq2, specifying `Gtype.Tx` as the condition of interest. 

```{r PCArlog3}
p.all <- plotPCA(rld, intgroup = c('Gtype.Tx'), ntop = 500)
p.all
```

Next, we'll save this plot to file.
```{r PCArlog1}
pdf(file = paste0(plotPath, 'PCAplot_rlog_', Comparison, '.pdf'), onefile = TRUE)
p.all
dev.off()
```


<details>
    <summary>*Click for code to customize PCA plots*</summary>
    We can generate a customized plot using the `ggplot2` package, including shape assignments for our treatment groups and color assignments for the individual samples.
```{r PCArlog4}
pdf(file = paste0(plotPath, 'PCAplot_Fancy_rlog_', Comparison, '.pdf'), onefile = TRUE)

#PCA plot for Rlog-Normalized counts for all samples
CombinatoricGroup <- factor(MetaInfo$Gtype.Tx)
SampleName <- factor(row.names(MetaInfo))


#Generate the PCA projections using the `plotPCA` function from DESeq2.
p.all <- plotPCA(rld, intgroup = c('Gtype.Tx'), ntop = 500)

gp <- ggplot(p.all$data, aes(x = PC1, y = PC2, color = SampleName, shape = CombinatoricGroup)) +
    scale_shape_manual(values=1:nlevels(CombinatoricGroup), name = 'Combinatoric Group') +
    geom_point(size=2) +
    ggtitle(label = as.character('All samples Rlog-Normalized')) +
    guides(colour=guide_legend(nrow=12, title = 'Sample'),
        legend.key = element_rect(size = 1),
        legend.key.size = unit(0, 'cm')) +
    xlab(p.all$labels[2]) + ylab(p.all$labels[1]) +
    theme(legend.margin = margin(t = 0, unit='mm'), plot.title = element_text(hjust = 0.5)) +
    theme_classic(base_size = 10)
plot(gp)
dev.off()
```
</details>



<details>
    <summary>*Click for example code for generating a ScreePlot*</summary>
     A screeplot is a way to visualize the variance explained by all principle components. 
     To generate a scree plot, the PCA results need to be used independently of plotting, such as described by [this statquest post](https://statquest.org/pca-clearly-explained/) and replicated below.
    ```{r ScreePlot}
    # generate PCA loadings
    pca <- prcomp(t(assay(rld)), scale=TRUE)

    ## get the scree information
    pca.var <- pca$sdev^2
    scree <- pca.var/sum(pca.var)
    barplot((scree[1:10]*100), main="Scree Plot", xlab="Principal Component", ylab="Percent Variation") 

    ```
    We can see that the majority (~65%) of the variance across our samples is explained by the first three principle components, giving us some additional confidence regarding the quality of our data.
    In these scree plot examples from BioTuring, the plot on the left fits what we would expect for a dataset with high signal from the experimental treatment, where the majority of the variance is explained by the first few principle components. The plot on the right illustrates a scenario where the variance is distributed across many components, which could be due to low signal from the experimental treatment, complex experimental design, or confounding factors.
image: ![](./images/proportion-of-variance-blog-horz.jpg)
</details>


#### Optional PCA plot for raw data
<details>
    <summary>*Click for code for PCA visualization of raw data*</summary>
    It can sometimes be useful to also generate a PCA plot for the raw data as well as the normalized data.
    ```{r PCAraw}
    pdf(file = paste0(plotPath, 'PCAplot_raw_', Comparison, '.pdf'), onefile = TRUE)
    #PCA for Raw counts for all samples
    RC <- SummarizedExperiment(log2(counts(dds, normalized = FALSE)), colData=colData(dds))
    p.RC <- plotPCA(DESeqTransform(RC), intgroup = 'Gtype.Tx')

    gpRC <- ggplot(p.RC$data, aes(x = PC1, y = PC2, color = SampleName, shape = CombinatoricGroup)) +     xlab(p.RC$labels[2]) + ylab(p.RC$labels[1]) + scale_shape_manual(values=1:nlevels(CombinatoricGroup), name = 'Combinatoric Group') + geom_point(size=2) + ggtitle(label = as.character('All samples Raw Data')) + theme(plot.title = element_text(hjust = 0.5)) + guides(colour=guide_legend(nrow=12, title = 'Sample'), legend.key = element_rect(size = 1), legend.key.size = unit(0, 'cm')) + theme_classic(base_size = 10) + theme(legend.margin=margin(t = 0, unit='mm'))
    plot(gpRC)
    
    dev.off()

    # embedd example of plot (rlog only)
    plot(gpRC)
    ```
    We see that there is less variance explained by PC1 and that the samples from the same group are not as well clustered for the raw data. Since this is prior to normalization, these differences are likely due to **technical** considerations like sequencing depth differences that are accounted for by the rlog normalization.   
</details>

**TODO: Add in summary from last workshop materials**



---


# Sources Used    
* HBC QC tutorial: https://hbctraining.github.io/DGE_workshop/lessons/03_DGE_QC_analysis.html    
* Detailed Heatmap tutorial from Galaxy: https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-heatmap2/tutorial.html   
* PCA Overview: https://blog.bioturing.com/2018/06/14/principal-component-analysis-explained-simply/     



```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
#Hidden code block to write out data for knitting
save.image(file = "rdata/RunningData.RData")
```


---

# Session Info:
```{r SessionInfo}
sessionInfo()
```

---


These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module09_SampleQCViz.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
