<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2021-08-08" />

<title>Day 2 - Module 10: DE Comparisons Results</title>

<script src="site_libs/header-attrs-2.8/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02_QC.html">Quality Control</a>
    </li>
    <li>
      <a href="Module02optional_Cutadapt_MultiQC.md">Optional: Cutadapt/MultiQC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03_Reference_Genomes.html">Reference Genomes</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module04_Alignment.html">Alignment/Quantification (STAR/RSEM)</a>
    </li>
    <li>
      <a href="Module05_MultiQC_and_Count_Matrix.html">Count matrix / Alignment QC</a>
    </li>
    <li>
      <a href="Module05optional_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">Introduction to DESeq2</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DESeq2 Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">Count Normalizations and DESeq2 model fitting</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample Visualizations for QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Comparisons and Gene Annotations</a>
    </li>
    <li>
      <a href="Module11_DEVisualizations.html">DE Visualizations and output files</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Day 2 - Module 10: DE Comparisons Results</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2021-08-08</h4>

</div>


<!--- Allow the page to be wider --->
<style>
    body .main-container {
        max-width: 1200px;
    }
</style>
<blockquote>
<h1 id="objectives">Objectives</h1>
<ul>
<li>Generate tables of DE results</li>
<li>Understand what a p-value represents.<br />
</li>
<li>Understand multiple hypothesis correction application and importance</li>
</ul>
</blockquote>
<div id="differential-expression-workflow" class="section level1" number="1">
<h1 number="1"><span class="header-section-number">1</span> Differential Expression Workflow</h1>
<p>Here we will finally test for differential expression between our comparisons of interest.</p>
<table>
<thead>
<tr class="header">
<th align="center">Step</th>
<th>Task</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">1</td>
<td>Experimental Design</td>
</tr>
<tr class="even">
<td align="center">2</td>
<td>Biological Samples / Library Preparation</td>
</tr>
<tr class="odd">
<td align="center">3</td>
<td>Sequence Reads</td>
</tr>
<tr class="even">
<td align="center">4</td>
<td>Assess Quality of Raw Reads</td>
</tr>
<tr class="odd">
<td align="center">5</td>
<td>Splice-aware Mapping to Genome</td>
</tr>
<tr class="even">
<td align="center">6</td>
<td>Count Reads Associated with Genes</td>
</tr>
<tr class="odd">
<td align="center">—</td>
<td>—</td>
</tr>
<tr class="even">
<td align="center">7</td>
<td>Organize project files locally</td>
</tr>
<tr class="odd">
<td align="center">8</td>
<td>Initialize DESeq2 and fit DESeq2 model</td>
</tr>
<tr class="even">
<td align="center">9</td>
<td>Assess expression variance within treatment groups</td>
</tr>
<tr class="odd">
<td align="center"><strong>10</strong></td>
<td><strong>Specify pairwise comparisons and test for differential expression</strong></td>
</tr>
<tr class="even">
<td align="center">11</td>
<td>Generate summary figures for comparisons</td>
</tr>
<tr class="odd">
<td align="center">12</td>
<td>Annotate differential expression result tables</td>
</tr>
</tbody>
</table>
<hr />
</div>
<div id="generating-de-results" class="section level1" number="2">
<h1 number="2"><span class="header-section-number">2</span> Generating DE results</h1>
<p>Now that we have reviewed the plots by sample and determined that our data passed our quality control checks, specifically that the patterns we observe are likely due to our experimental treatments over technical or other cofounding factors.</p>
<p>This illustration from the HCB training materials illustrates the basis of the differential expression procedure, where our goal is to compare the distribution of an expressed gene across samples in each treatment groups. <img src="images/de_theory.png" /> <em>Image credit: Paul Pavlidis, UBC</em></p>
<p>Only where the distributions of each group are sufficiently seperated will a gene be considered differentially expressed. This is where having sufficient replicates to overcome within group variance is important, as the more replicates we have in each group the better we can determine the distributions of expression for each group.</p>
<div id="deseq2-statistical-testing" class="section level2" number="2.1">
<h2 number="2.1"><span class="header-section-number">2.1</span> DESeq2 statistical testing</h2>
<p>We have already fit our DESeq2 model, specifing our model as <code>~ Gtype.Tx</code> and our next step is to identify genes with significantly different expression between our contrasts of interest. To determine significance, a statistical test is required.</p>
<p>The first step for any statistical test is to define the <em>null hypothesis</em>. In this case, the null hypothesis would be that there is no difference in the expression of a gene between two groups of samples, such as illustrated at the bottom of the first figure in this module. The next step is to use a statistical test to determine if, based on the observed data, the null hypothesis can be rejected.</p>
<p>To do this, <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#theory-behind-deseq2">DESeq2 applies the Wald test</a> to compare two groups. A Wald test statistic is computed as well as the probability that the observed value or more extreme test statistic would be observed. This probability is called the p-value of the test. If the p-value is smaller than a pre-defined threshold, we would reject the null hypothesis and state that there is evidence against the null, i.e. the gene is differentially expressed. However, if the p-value is larger than our threshold, we would <em>fail to reject</em> the null hypothesis, meaning that we lack evidence that the expression of this gene is different <em>NOT</em> that we have evidence that the expression is indeed the same in both groups.</p>
<p>For a more detailed overview of the statistical comparisons , please refer to <a href="https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html">this HBC tutorial</a> or the <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#theory-behind-deseq2">DESeq2 vignette</a>.</p>
<p><strong>TODO: Change to dropdown/optional content</strong> ## Dispersion estimates</p>
<p>Since we’ve already run our <code>DESeq</code> analysis, we can visualize the DESeq2 normalization results for our data, which centers on shrinking the variance across all genes to better fit the expected spread at a given expression level by plotting the <strong>dispersion estimates</strong> with the <code>plotDispEsts</code> function.</p>
<pre class="r"><code>plotDispEsts(dds)</code></pre>
<p><img src="Module10_DEComparisons_files/figure-html/CheckDispersions-1.png" width="672" /></p>
<p>We can see the raw data plotted in black, the fitted (or expected) dispersion in red, and the normalized data with scaled variance in blue. Since we have fairly small sample sizes for each condition, we see shrinkage for many genes but a reasonable correlation between the expression level and dispersions.</p>
<p>This <a href="https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html">HBC tutorial</a> has a more detailed overview of estimating size factors, estimating gene dispersion, and the shrinkage procedure, as well as examples of concerning dispersion plots that may suggest reassessing quality of the experimental data.</p>
</div>
<div id="results-table-output" class="section level2" number="2.2">
<h2 number="2.2"><span class="header-section-number">2.2</span> Results table output</h2>
<p><strong>TODO: Add image or hidden code block to show results example - perhaps from DESeq2 documentation so more generic? </strong></p>
<p>In the results table that we’ll generate the row name are gene symbols and there are <a href="https://support.illumina.com/help/BS_App_RNASeq_DE_OLH_1000000071939/Content/Source/Informatics/Apps/DESeq2ResultFile_swBS.htm#:~:text=baseMean%E2%80%94The%20average%20of%20the,factors%2C%20taken%20over%20all%20samples.&amp;text=log2FoldChange%E2%80%93The%20effect%20size%20estimate,the%20comparison%20and%20control%20groups">six columns of values</a>.</p>
<p>The first column, ‘baseMean’ is the average of the normalized count values, dividing by size factors, taken over all samples, and can be interpreted as the relative expression level of that gene across all samples.</p>
<p>The second column, ‘log2FoldChange’, is the ratio of the expression of the numerator group (<code>ko.Tx</code>) over the denominator group (<code>wt.Tx</code>). If the value is positive, that means the expression of that gene is greater across the <code>ko.Tx</code> samples than across the <code>wt.Tx</code> samples. If the value is negative, that means the expression of that gene is greater across the <code>wt.Tx</code> samples. The third column, ‘lfcSE’ is the standard error for the log2 fold change estimate.</p>
<p>The fourth column, ‘stat’, is the calculated Wald statistic for that gene, while the fifth column ‘pvalue’ is the <em>nominal</em> significance for that gene.</p>
<div id="multiple-hypothesis-testing-and-fdr-correction" class="section level3" number="2.2.1">
<h3 number="2.2.1"><span class="header-section-number">2.2.1</span> Multiple hypothesis testing and FDR correction</h3>
<p>The sixth column, ‘padj’, is the <em>adjusted p-value</em> and is what we use for determining significantly differently expressed genes. Why do we use values from this column instead of the ‘pvalue’ column?</p>
<p>Each p-value is the result of a single test for a single gene. The more genes we test, the greater chance we have of seeing a significant results. This is the multiple testing problem. If we used the p-value directly from the Wald test with a significance cut-off of p &lt; 0.05, that means there is a 5% chance it is a false positives. So if we are testing 20,000 genes for differential expression, we would expect to see ~1,000 significant genes just by chance. This is problematic because we would need to sift through our “significant” genes to identify which ones are true positives.</p>
<p>DESeq2 reduces the number of genes that will be tested by removing genes with low number of counts and outlier samples (gene-level QC, see note below). However, we need to correct for <a href="https://multithreaded.stitchfix.com/blog/2015/10/15/multiple-hypothesis-testing/">multiple hypothesis testing</a> to reduce the number of false positives, and while there are a few common approaches, the default method is False Discovery Rate(FDR)/Benjamini-Hochberg correction which is symbolized as ‘BH’ in DESeq2. <a href="https://rss.onlinelibrary.wiley.com/doi/10.1111/j.2517-6161.1995.tb02031.x">Benjamini and Hochberg (1995)</a> defined the concept of FDR and created an algorithm to control it below a specified level. An interpretation of the BH method is implemented in DESeq2 in which genes are ranked by p-value, then each ranked p-value is multiplied by the number of total tests divided by rank.</p>
<blockquote>
<p><strong>Note</strong>: From the <code>results</code> function help page and <a href="https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html">HBC tutorial that includes overview of multiple hypothesis correction</a>, we can change the multiple hypothesis correction method to an alternative option using the <code>pAdjustMethod =</code> argument.</p>
</blockquote>
<p>The default FDR rate cutoff for DESeq2 is <code>alpha = 0.05</code>. By setting the cutoff to &lt; 0.05, we expect that the proportion of false positives amongst our differentially expressed genes is now controlled to 5%. For example, if we call 500 genes as differentially expressed with this FDR cutoff, we expect only 25 of them to be false positives. DESeq2 vignette’s includes a <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#independent-filtering-and-multiple-testing">further discussion of filtering and multiple testing</a></p>
<blockquote>
<p><strong>Note on ‘padj’ values set to NA</strong><br />
As discussed in the <a href="https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html">HBC tutorial</a> as well as the <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#i-want-to-benchmark-deseq2-comparing-to-other-de-tools.">DESeq2 vignette</a><br />
* If within a row, all samples have zero counts, the baseMean column will be zero, and the log2 fold change estimates, p-value and adjusted p-value will all be set to NA.<br />
* If a row contains a sample with an extreme count outlier then the p-value and adjusted p-value will be set to NA. These outlier counts are detected by Cook’s distance.<br />
* If a row is filtered by automatic independent filtering, for having a low mean normalized count, then only the adjusted p-value will be set to NA.</p>
</blockquote>
</div>
</div>
<div id="generating-results" class="section level2" number="2.3">
<h2 number="2.3"><span class="header-section-number">2.3</span> Generating results</h2>
<p>We can check what comparisons were automatically generated during fitting using the <code>resultsNames()</code>.</p>
<pre class="r"><code>resultsNames(dds)</code></pre>
<pre><code>## [1] &quot;Intercept&quot;                    &quot;Gtype.Tx_ko.Tx_vs_wt.Tx&quot;      &quot;Gtype.Tx_ko.control_vs_wt.Tx&quot;
## [4] &quot;Gtype.Tx_wt.control_vs_wt.Tx&quot;</code></pre>
<p>Since we are interested in comparing each knockout versus its corresponding wild-type control, only one of the automatically generated comparisons is relevant. We can pull the <code>Tx_ko.Tx_vs_wt.Tx</code> comparison results using the <code>results</code> function and assign those result to a new object.</p>
<pre class="r"><code>Comparison &lt;- &quot;Gtype.Tx_ko.Tx_vs_wt.Tx&quot;
res_Tx &lt;- results(dds, name=Comparison)</code></pre>
</div>
<div id="how-to-generate-additional-contrasts" class="section level2" number="2.4">
<h2 number="2.4"><span class="header-section-number">2.4</span> How to generate additional contrasts</h2>
<p>If there are comparisons that are not included in the <code>resultsName</code>, since our dds object already has the fitted data we can generate Wald test results by specifying those comparisons as an additional arguement to the results function.</p>
<pre class="r"><code>?results</code></pre>
<p>As the function description specifies, we need to provide a list of three elements: the name of the factor in the model design, the name of the numerator for the fold-change, and the name of denominator.</p>
<pre class="r"><code>res_WT &lt;- results(dds, contrast=c(&quot;Gtype.Tx&quot;, &quot;ko.control&quot;, &quot;wt.control&quot;)) 
head(res_WT)</code></pre>
<pre><code>## log2 fold change (MLE): Gtype.Tx ko.control vs wt.control 
## Wald test p-value: Gtype.Tx ko.control vs wt.control 
## DataFrame with 6 rows and 6 columns
##                      baseMean log2FoldChange     lfcSE      stat      pvalue       padj
##                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;  &lt;numeric&gt;
## ENSMUSG00000000001 6255.78733    -0.01370173 0.0931866 -0.147035 0.883104082 0.96100310
## ENSMUSG00000000028 1337.98982     0.52267884 0.1359753  3.843924 0.000121082 0.00391537
## ENSMUSG00000000031    3.77363    -1.15621509 1.6006132 -0.722358 0.470074665         NA
## ENSMUSG00000000037   27.56347    -0.27921011 0.4764764 -0.585989 0.557882649 0.80447495
## ENSMUSG00000000049    2.25696     4.01133495 1.8890848  2.123428 0.033718026         NA
## ENSMUSG00000000056 2194.17848    -0.00821319 0.1773220 -0.046318 0.963056826 0.98888983</code></pre>
<p>Last, we’ll at our own results table.</p>
<pre class="r"><code>head(res_Tx)</code></pre>
<pre><code>## log2 fold change (MLE): Gtype.Tx ko.Tx vs wt.Tx 
## Wald test p-value: Gtype.Tx ko.Tx vs wt.Tx 
## DataFrame with 6 rows and 6 columns
##                      baseMean log2FoldChange     lfcSE      stat      pvalue        padj
##                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;
## ENSMUSG00000000001 6255.78733       0.079621 0.0934295  0.852204 3.94101e-01 6.64152e-01
## ENSMUSG00000000028 1337.98982       0.226230 0.1161800  1.947239 5.15061e-02 2.15863e-01
## ENSMUSG00000000031    3.77363       0.579460 1.1621131  0.498626 6.18043e-01          NA
## ENSMUSG00000000037   27.56347      -0.555188 0.4306035 -1.289324 1.97285e-01 4.66009e-01
## ENSMUSG00000000049    2.25696       0.187564 1.6951093  0.110650 9.11894e-01          NA
## ENSMUSG00000000056 2194.17848      -1.306315 0.1778575 -7.344731 2.06174e-13 3.20181e-11</code></pre>
<p>Now that we’ve generated our differential comparisons and have an understanding of our results, including multiple hypothesis correction, we can proceed with generating summary figures and tables for our differential expression analysis.</p>
<hr />
</div>
</div>
<div id="sources-used" class="section level1" number="3">
<h1 number="3"><span class="header-section-number">3</span> Sources Used</h1>
<ul>
<li>HBC DGE training module, part 1: <a href="https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html" class="uri">https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html</a></li>
<li>HBC DGE training module, part 2: <a href="https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html" class="uri">https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html</a></li>
<li>DESeq2 vignette: <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis" class="uri">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis</a></li>
</ul>
<hr />
<p><img src="images/significant_xkcd.png" /></p>
<hr />
</div>
<div id="session-info" class="section level1" number="4">
<h1 number="4"><span class="header-section-number">4</span> Session Info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 11.5.1
## 
## Matrix products: default
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] dplyr_1.0.6                 data.table_1.14.0           RColorBrewer_1.1-2         
##  [4] pheatmap_1.0.12             ggrepel_0.9.1               tidyr_1.1.3                
##  [7] ggplot2_3.3.4               DESeq2_1.32.0               SummarizedExperiment_1.22.0
## [10] Biobase_2.30.0              MatrixGenerics_1.4.0        matrixStats_0.58.0         
## [13] GenomicRanges_1.44.0        GenomeInfoDb_1.28.0         IRanges_2.26.0             
## [16] S4Vectors_0.30.0            BiocGenerics_0.38.0         rmarkdown_2.8              
## 
## loaded via a namespace (and not attached):
##  [1] sass_0.4.0             bit64_4.0.5            jsonlite_1.7.2         splines_4.1.0         
##  [5] bslib_0.2.5.1          assertthat_0.2.1       highr_0.9              blob_1.2.1            
##  [9] GenomeInfoDbData_1.2.6 yaml_2.2.1             pillar_1.6.1           RSQLite_2.2.7         
## [13] lattice_0.20-44        glue_1.4.2             digest_0.6.27          XVector_0.32.0        
## [17] colorspace_2.0-1       htmltools_0.5.1.1      Matrix_1.3-4           XML_3.99-0.6          
## [21] pkgconfig_2.0.3        genefilter_1.74.0      zlibbioc_1.16.0        purrr_0.3.4           
## [25] xtable_1.8-4           scales_1.1.1           BiocParallel_1.26.0    tibble_3.1.2          
## [29] annotate_1.48.0        farver_2.1.0           generics_0.1.0         ellipsis_0.3.2        
## [33] withr_2.4.2            cachem_1.0.5           survival_3.2-11        magrittr_2.0.1        
## [37] crayon_1.4.1           memoise_2.0.0          evaluate_0.14          fansi_0.5.0           
## [41] tools_4.1.0            lifecycle_1.0.0        stringr_1.4.0          locfit_1.5-9.4        
## [45] munsell_0.5.0          DelayedArray_0.18.0    AnnotationDbi_1.32.3   compiler_4.1.0        
## [49] jquerylib_0.1.4        rlang_0.4.11           grid_4.1.0             RCurl_1.98-1.3        
## [53] labeling_0.4.2         bitops_1.0-7           gtable_0.3.0           DBI_1.1.1             
## [57] R6_2.5.0               knitr_1.33             fastmap_1.1.0          bit_4.0.4             
## [61] utf8_1.2.1             stringi_1.6.2          Rcpp_1.0.6             vctrs_0.3.8           
## [65] geneplotter_1.70.0     tidyselect_1.1.1       xfun_0.23</code></pre>
<hr />
<p>These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the <a href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.</p>
</div>

<div id="rmd-source-code">---
title: "Day 2 - Module 10: DE Comparisons Results"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: true
            fig_caption: true
            markdown: GFM
            code_download: true
---

<!--- Allow the page to be wider --->
<style>
    body .main-container {
        max-width: 1200px;
    }
</style>

> # Objectives 
> * Generate tables of DE results
> * Understand what a p-value represents.   
> * Understand multiple hypothesis correction application and importance    


```{r Modules, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
library(tidyr)
library(dplyr)
load("rdata/RunningData.RData")
```

# Differential Expression Workflow

Here we will finally test for differential expression between our comparisons of interest.

| Step | Task |
| :--: | ---- |
| 1 | Experimental Design |
| 2 | Biological Samples / Library Preparation |
| 3 | Sequence Reads |
| 4 | Assess Quality of Raw Reads |
| 5 | Splice-aware Mapping to Genome |
| 6 | Count Reads Associated with Genes |
| --- | --- |
| 7 | Organize project files locally |
| 8 | Initialize DESeq2 and fit DESeq2 model |
| 9 | Assess expression variance within treatment groups |
| **10** | **Specify pairwise comparisons and test for differential expression** |
| 11 | Generate summary figures for comparisons |
| 12 | Annotate differential expression result tables |

---

# Generating DE results

Now that we have reviewed the plots by sample and determined that our data passed our quality control checks, specifically that the patterns we observe are likely due to our experimental treatments over technical or other cofounding factors.

This illustration from the HCB training materials illustrates the basis of the differential expression procedure, where our goal is to compare the distribution of an expressed gene across samples in each treatment groups.
![](./images/de_theory.png)
*Image credit: Paul Pavlidis, UBC*

Only where the distributions of each group are sufficiently seperated will a gene be considered differentially expressed. This is where having sufficient replicates to overcome within group variance is important, as the more replicates we have in each group the better we can determine the distributions of expression for each group.

## DESeq2 statistical testing

We have already fit our DESeq2 model, specifing our model as `~ Gtype.Tx` and our next step is to identify genes with significantly different expression between our contrasts of interest. To determine significance, a statistical test is required.

The first step for any statistical test is to define the *null hypothesis*. In this case, the null hypothesis would be that there is no difference in the expression of a gene between two groups of samples, such as illustrated at the bottom of the first figure in this module. The next step is to use a statistical test to determine if, based on the observed data, the null hypothesis can be rejected.

To do this, [DESeq2 applies the Wald test](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#theory-behind-deseq2) to compare two groups. A Wald test statistic is computed as well as the probability that the observed value or more extreme test statistic would be observed. This probability is called the p-value of the test. If the p-value is smaller than a pre-defined threshold, we would reject the null hypothesis and state that there is evidence against the null, i.e. the gene is differentially expressed. However, if the p-value is larger than our threshold, we would *fail to reject* the null hypothesis, meaning that we lack evidence that the expression of this gene is different *NOT* that we have evidence that the expression is indeed the same in both groups. 


For a more detailed overview of the statistical comparisons , please refer to [this HBC tutorial](https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html) or the [DESeq2 vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#theory-behind-deseq2).


**TODO: Change to dropdown/optional content**
## Dispersion estimates

Since we've already run our `DESeq` analysis, we can visualize the DESeq2 normalization results for our data, which centers on shrinking the variance across all genes to better fit the expected spread at a given expression level by plotting the **dispersion estimates** with the `plotDispEsts` function.

```{r CheckDispersions}
plotDispEsts(dds)
```

We can see the raw data plotted in black, the fitted (or expected) dispersion in red, and the normalized data with scaled variance in blue. Since we have fairly small sample sizes for each condition, we see shrinkage for many genes but a reasonable correlation between the expression level and dispersions.

This [HBC tutorial](https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html) has a more detailed overview of estimating size factors, estimating gene dispersion, and the shrinkage procedure, as well as examples of concerning dispersion plots that may suggest reassessing quality of the experimental data. 



## Results table output

**TODO: Add image or hidden code block to show results example - perhaps from DESeq2 documentation so more generic? **

In the results table that we'll generate the row name are gene symbols and there are [six columns of values](https://support.illumina.com/help/BS_App_RNASeq_DE_OLH_1000000071939/Content/Source/Informatics/Apps/DESeq2ResultFile_swBS.htm#:~:text=baseMean%E2%80%94The%20average%20of%20the,factors%2C%20taken%20over%20all%20samples.&text=log2FoldChange%E2%80%93The%20effect%20size%20estimate,the%20comparison%20and%20control%20groups). 

The first column, 'baseMean' is the average of the normalized count values, dividing by size factors, taken over all samples, and can be interpreted as the relative expression level of that gene across all samples. 

The second column, 'log2FoldChange', is the ratio of the expression of the numerator group (`ko.Tx`) over the denominator group (`wt.Tx`). If the value is positive, that means the expression of that gene is greater across the `ko.Tx` samples than across the `wt.Tx` samples. If the value is negative, that means the expression of that gene is greater across the `wt.Tx` samples. The third column, 'lfcSE' is the standard error for the log2 fold change estimate. 

The fourth column, 'stat', is the calculated Wald statistic for that gene, while the fifth column 'pvalue' is the *nominal* significance for that gene.

### Multiple hypothesis testing and FDR correction

The sixth column, 'padj', is the *adjusted p-value* and is what we use for determining significantly differently expressed genes. Why do we use values from this column instead of the 'pvalue' column?

Each p-value is the result of a single test for a single gene. The more genes we test, the greater chance we have of seeing a significant results. This is the multiple testing problem. If we used the p-value directly from the Wald test with a significance cut-off of p < 0.05, that means there is a 5% chance it is a false positives. So if we are testing 20,000 genes for differential expression, we would expect to see ~1,000 significant genes just by chance. This is problematic because we would need to sift through our “significant” genes to identify which ones are true positives.

DESeq2 reduces the number of genes that will be tested by removing genes with low number of counts and outlier samples (gene-level QC, see note below). However, we need to correct for [multiple hypothesis testing](https://multithreaded.stitchfix.com/blog/2015/10/15/multiple-hypothesis-testing/) to reduce the number of false positives, and while there are a few common approaches, the default method is False Discovery Rate(FDR)/Benjamini-Hochberg correction which is symbolized as 'BH' in DESeq2. [Benjamini and Hochberg (1995)](https://rss.onlinelibrary.wiley.com/doi/10.1111/j.2517-6161.1995.tb02031.x) defined the concept of FDR and created an algorithm to control it below a specified level. An interpretation of the BH method is implemented in DESeq2 in which genes are ranked by p-value, then each ranked p-value is multiplied by the number of total tests divided by rank.

> **Note**: From the `results` function help page and [HBC tutorial that includes overview of multiple hypothesis correction](https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html), we can change the multiple hypothesis correction method to an alternative option using the `pAdjustMethod =` argument.

The default FDR rate cutoff for DESeq2 is `alpha = 0.05`. By setting the cutoff to < 0.05, we expect that the proportion of false positives amongst our differentially expressed genes is now controlled to 5%. For example, if we call 500 genes as differentially expressed with this FDR cutoff, we expect only 25 of them to be false positives. DESeq2 vignette's includes a [further discussion of filtering and multiple testing](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#independent-filtering-and-multiple-testing)

> **Note on 'padj' values set to NA**    
> As discussed in the [HBC tutorial](https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html) as well as the [DESeq2 vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#i-want-to-benchmark-deseq2-comparing-to-other-de-tools.)    
> * If within a row, all samples have zero counts, the baseMean column will be zero, and the log2 fold change estimates, p-value and adjusted p-value will all be set to NA.    
> * If a row contains a sample with an extreme count outlier then the p-value and adjusted p-value will be set to NA. These outlier counts are detected by Cook’s distance.     
> * If a row is filtered by automatic independent filtering, for having a low mean normalized count, then only the adjusted p-value will be set to NA.   

## Generating results

We can check what comparisons were automatically generated during fitting using the `resultsNames()`. 
```{r Results1}
resultsNames(dds)
```

Since we are interested in comparing each knockout versus its corresponding wild-type control, only one of the automatically generated comparisons is relevant. We can pull the `Tx_ko.Tx_vs_wt.Tx` comparison results using the `results` function and assign those result to a new object.
```{r SetsResult}
Comparison <- "Gtype.Tx_ko.Tx_vs_wt.Tx"
res_Tx <- results(dds, name=Comparison)
```

## How to generate additional contrasts

If there are comparisons that are not included in the `resultsName`, since our dds object already has the fitted data we can generate Wald test results by specifying those comparisons as an additional arguement to the results function. 
```{r ResultsFunction}
?results
```

As the function description specifies, we need to provide a list of three elements: the name of the factor in the model design, the name of the numerator for the fold-change, and the name of denominator. 
```{r AdditionalComparisons}
res_WT <- results(dds, contrast=c("Gtype.Tx", "ko.control", "wt.control")) 
head(res_WT)
```


Last, we'll at our own results table.
```{r ResultsKD}
head(res_Tx)
```

Now that we've generated our differential comparisons and have an understanding of our results, including multiple hypothesis correction, we can proceed with generating summary figures and tables for our differential expression analysis.

---


# Sources Used
* HBC DGE training module, part 1: https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html
* HBC DGE training module, part 2: https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html
* DESeq2 vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis

---

![](./images/significant_xkcd.png)


```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
#Hidden code block to write out data for knitting
save.image(file = "rdata/RunningData.RData")
```

---

# Session Info
```{r SessionInfo}
sessionInfo()
```

---


These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module10_DEComparisons.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
